include/boost/json/impl/pointer.ipp

```diff
@@ -0,0 +1,329 @@
+//
+// Copyright (c) 2021 Dmitry Arkhipov (grisumbras@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/json
+//
+
+#ifndef BOOST_JSON_IMPL_POINTER_IPP
+#define BOOST_JSON_IMPL_POINTER_IPP
+
+#include <boost/json/pointer.hpp>
+
+BOOST_JSON_NS_BEGIN
+
+namespace detail {
+
+
+class pointer_token
+{
+public:
+    class iterator;
+
+    pointer_token(
+        string_view::const_iterator b,
+        string_view::const_iterator e) noexcept
+        : b_(b), e_(e)
+    {
+    }
+
+    iterator begin() const noexcept;
+    iterator end() const noexcept;
+
+private:
+    string_view::const_iterator b_;
+    string_view::const_iterator e_;
+};
+
+class pointer_token::iterator
+{
+public:
+    explicit iterator(string_view::const_iterator base) noexcept
+        : base_(base)
+    {
+    }
+
+    char operator*() const noexcept
+    {
+        switch (char c = *base_)
+        {
+        case '~':
+            c = base_[1];
+            if ( '0' == c )
+            {
+                return '~';
+            }
+            else
+            {
+                BOOST_ASSERT('1' == c);
+                return '/';
+            }
+        default:
+            return c;
+        }
+    }
+
+    iterator& operator++() noexcept
+    {
+        if ( '~' == *base_ )
+        {
+            base_ += 2;
+        }
+        else
+        {
+            ++base_;
+        }
+        return *this;
+    }
+
+    string_view::const_iterator base() const noexcept
+    {
+        return base_;
+    }
+
+private:
+    string_view::const_iterator base_;
+};
+
+bool operator==(pointer_token::iterator l, pointer_token::iterator r) noexcept
+{
+    return l.base() == r.base();
+}
+
+bool operator!=(pointer_token::iterator l, pointer_token::iterator r) noexcept
+{
+    return l.base() != r.base();
+}
+
+pointer_token::iterator pointer_token::begin() const noexcept
+{
+    return iterator(b_);
+}
+
+pointer_token::iterator pointer_token::end() const noexcept
+{
+    return iterator(e_);
+}
+
+bool operator==(pointer_token token, string_view sv) noexcept
+{
+    auto t_b = token.begin();
+    auto const t_e = token.end();
+    auto s_b = sv.begin();
+    auto const s_e = sv.end();
+    for (; t_b != t_e && s_b != s_e; ++t_b, ++s_b)
+    {
+        if (*t_b != *s_b)
+        {
+            return false;
+        }
+    }
```

Comment:
Too many statements on the same line, which confuses coverage and makes setting breakpoints more difficult. I would rewrite it to use fewer compound expressions:
```
bool
operator==(
    pointer_token token,
    string_view sv) noexcept
{
    auto n = sv.size();
    auto s = sv.data();
    auto p = token.begin();
    auto const end = token.end();
    while(n--)
    {
        if(p == end)
            return false;
        if(*p != *s)
            return false;
        ++p;
        ++s;
    }
    if(p != end)
        return false;
    return true;
}
```